### PR TITLE
ext

### PR DESCRIPTION
**Pull request by kanake10**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [ ] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [ ] the article explains *why* and not only *how*
- [ ] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [ ] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [ ] the text is conversational ("we" instead of "you", easy language)
- [ ] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [ ] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [ ] the text uses active voice instead of passive voice
- [ ] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [ ] file names and variables are highlighted as `code`

## Structure
- [ ] the article has a short intro before the first headline
- [ ] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [ ] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [ ] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [ ] code examples are introduced with a ":" and explained below the code example
- [ ] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

File: content/blog/2023/2023-12-17-extension-functions-in-kotlin.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+title: "Exploring the Power of extension Functions in Kotlin"
+categories: ["Kotlin"]
+date: 2023-12-13 00:00:00 +1100
+authors: [ezra]
+excerpt: "In this tutorial, we'll discuss what extension functions are in kotlin and their usecases"
+image: images/stock/0104-on-off-1200x628-branded.jpg
+url: extension functions in kotlin
+---
+
+One of the Kotlin's standout features is extension functions, a mechanism that empowers developers to enhance existing classes without modifying their source code. In this blog post, we will embark on a journey to explore the fascinating realm of Kotlin extension functions, understanding their syntax, exploring use cases and recognizing their impact on code maintainability and readability.
```
Comment (thombergs) [inline content/blog/2023/2023-12-17-extension-functions-in-kotlin.md]: I took the liberty and dampened the language here a bit ... readers want to learn about extension functions, not "embark on a journey to explore the fascinating realm of Kotlin extension functions" ... :)

File: content/blog/2023/2023-12-17-extension-functions-in-kotlin.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+## Advantages of Extension Functions
+
+### Code Organization and Readability
+
+Extension functions contribute to better code organization by grouping related functionality together. This is particularly beneficial when working with large codebases, as it helps maintain a clear and modular structure. By encapsulating related operations in extension functions, developers can quickly locate and understand the purpose of specific functionality.
+
+Consider a scenario where we need to manipulate dates in various ways throughout our codebase. Instead of scattering date-related functions across different classes or files, we can create a `DateUtils` class and define extension functions for the Date class within it. This centralizes date-related operations, improving code readability and maintainability.
+
+```kotlin
+class DateUtils
```
Comment (thombergs) [inline content/blog/2023/2023-12-17-extension-functions-in-kotlin.md:41]: Is this valid Kotlin code? The body of a class in Kotlin needs braces (`{}`), too, doesn't it? Also, extension functions don't need to be defined inside of a class, do they?

File: content/blog/2023/2023-12-17-extension-functions-in-kotlin.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+
+## Best Practices for Using Extension Functions
+While extension functions offer a powerful tool for enhancing code, it's essential to follow some best practices to ensure their effective and maintainable usage.
+
+### Avoid Overuse
+While extension functions can greatly improve code readability and organization, excessive use can lead to confusion and code bloat. It's crucial to strike a balance and reserve extension functions for cases where they genuinely improve the API and maintainability of the code.
+
+### Be Mindful of Scope
+Extension functions are in scope wherever their containing package is imported. Be cautious about naming clashes to avoid unintentional overrides or confusion. Choose names that are descriptive and unlikely to conflict with existing or future API additions.
```
Comment (thombergs) [inline content/blog/2023/2023-12-17-extension-functions-in-kotlin.md:217]: The scope issue hasn't been addressed in the article until now. Can you add a section to the "Understanding Extension Functions" part that explains this? Or explain it with a short code example here in this section. How do I "import" an extension function? How can I influence the scope of an extension function?

From this section alone, I don't understand how it works.

**Review by thombergs — CHANGES_REQUESTED** [review]
Left some comments and questions and pushed some minor edits.

File: content/blog/2023/2023-12-17-extension-functions-in-kotlin.md (hunk 2 above)
Comment (kanake10) [reply content/blog/2023/2023-12-17-extension-functions-in-kotlin.md:41]: this was a typo and yes extension functions don't need to be defined inside a class

File: content/blog/2023/2023-12-17-extension-functions-in-kotlin.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+// File: Main.kt
+import com.example.util.customExtensionFunction
+
+fun main() {
+    val myString = "hello, world!"
+    val result = myString.customExtensionFunction()
+    println(result)
+}
+```
+In the `Main.kt` file, we import the `customExtensionFunction` from the `com.example.util` package and apply it to a string. The extension function is in scope wherever the package is imported.
```
Comment (thombergs) [inline content/blog/2023/2023-12-17-extension-functions-in-kotlin.md:240]: Sorry, but I still don't understand what you mean with "importing a package". We're importing the function here, not a package. Is it possible to import all extension functions inside a certain package in Kotlin? In any case, this section needs to be a bit clearer.

Comment (kanake10) [reply content/blog/2023/2023-12-17-extension-functions-in-kotlin.md:240]: The package in this case is simply the directory where we created our extension function.

Comment (thombergs) [reply content/blog/2023/2023-12-17-extension-functions-in-kotlin.md:240]: I changed "...wherever the package is imported" to " ... wherever the function is imported", then. Details matter :)

**Pull request merged**